packages/core/src/service/services/product.service.ts

```diff
@@ -84,16 +87,23 @@ export class ProductService {
     }
 
     async findOne(ctx: RequestContext, productId: ID): Promise<Translated<Product> | undefined> {
-        const product = await this.connection.findOneInChannel(ctx, Product, productId, ctx.channelId, {
-            relations: this.relations,
-            where: {
-                deletedAt: null,
-            },
-        });
-        if (!product) {
-            return;
-        }
-        return translateDeep(product, ctx.languageCode, ['facetValues', ['facetValues', 'facet']]);
+        const qb = this.connection.getRepository(ctx, Product).createQueryBuilder('product');
+        FindOptionsUtils.applyFindManyOptionsOrConditionsToQueryBuilder(qb, { relations: this.relations });
+        // tslint:disable-next-line:no-non-null-assertion
+        FindOptionsUtils.joinEagerRelations(qb, qb.alias, qb.expressionMap.mainAlias!.metadata);
+        return qb
+            .leftJoin('product.variants', 'variant')
```

Comment:
I'm wondering about any potential performance implications of this part. Imagine that a Product has 100 variants. To look up that product we then need to join 100 * (# channels) rows to check whether the Product is available on the current Channel.

In the original proposal we discussed that edge-case where a Product is created without any variants, and it therefore belongs to no Channel. One of my proposals was to keep a reference to the Channels on the Product. I'm now wondering whether it is worthwhile to keep it for reasons of performance and also developer ergonomics.

Of course this perf thing is just a suspicion. It needs to be tested. I'll put together a load test so we can have some objective idea about whether or not it really matters to do it this way.